# Cleaning data with `pandas`

```{admonition} Questions
:class: questions
- How can we deal with imperfect data?
- What is tidy data?
```
```{admonition} Objectives
:class: objectives
- Clean a dataset using `pandas`
```

In the previous section we learned how to read data in with `pandas` and how to access the data within a dataframe. Here we will be looking at how to "clean" our data while still staying within the bounds of reproducible science.

```{admonition} Exercise: Read in the data
:class: practice
If you haven't already, import `pandas` and read the exoplanet data into a dataframe. I'm calling the dataframe `exo_df`, the same as in the last section. If you're reading in the data for the first time now rename the first column to `name` rather than `# name`. You can check the [previous section](01_intro_pandas.ipynb) if you need a refresher on how to do this. Check that the data has been read correctly by looking at your dataframe.
```

In [2]:
import pandas as pd

In [3]:
exo_df = pd.read_csv("./data/Exoplanet_catalog_2019.csv")

In [4]:
exo_df.rename(columns={'# name':'name'}, inplace=True)

Let's take a quick look at our data again. One way to do this is with the `head()` function. This lets us display a certain number of rows. `head()` will display the first 10 rows by default, but you can pass it a number in the brackets if you'd like to see more or less. I want to see the first 20 rows:

In [6]:
exo_df.head(20)

name  orbital_period  semi_major_axis  eccentricity  discovered  \
0         11 Com b       326.03000          1.29000        0.2310      2008.0   
1         11 UMi b       516.22000          1.54000        0.0800      2009.0   
2         14 And b       185.84000          0.83000        0.0000      2008.0   
3         14 Her b      1773.40000          2.77000        0.3690      2002.0   
4       16 Cyg B b       799.50000          1.68000        0.6890      1996.0   
5         18 Del b       993.30000          2.60000        0.0800      2008.0   
6   1SWASP J1407 b      3725.00000          3.90000           NaN      2012.0   
7         24 Boo b        30.35060          0.19000        0.0420      2018.0   
8         24 Sex b       452.80000          1.33300        0.0900      2010.0   
9         24 Sex c       883.00000          2.08000        0.2900      2010.0   
10      30 Ari B b       335.10000          0.99500        0.2890      2009.0   
11        38 Vir b       825.90000          1.82000        0.0300      2016.0   
12         4 Uma b       269.30000          0.87000        0.4320      2007.0   
13        42 Dra b       479.10000          1.19000        0.3800      2009.0   
14        47 Uma b      1078.00000          2.10000        0.0320      1996.0   
15        47 Uma c      2391.00000          3.60000        0.0980      2001.0   
16        47 Uma d     14002.00000         11.60000        0.1600      2010.0   
17        51 Peg b         4.23080          0.05200        0.0069      1995.0   
18        55 Cnc b        14.65314          0.11339        0.0023      1996.0   
19        55 Cnc c        44.37300          0.23735        0.0720      2002.0   

     detection_type     star_name  star_distance  star_metallicity  star_mass  \
0   Radial Velocity        11 Com       110.6000            -0.350      2.700   
1   Radial Velocity        11 UMi       119.5000             0.040      1.800   
2   Radial Velocity        14 And        76.4000            -0.240      2.200   
3   Radial Velocity        14 Her        18.1000             0.430      0.900   
4   Radial Velocity      16 Cyg B        21.4100             0.080      1.010   
5   Radial Velocity        18 Del        73.1000            -0.052      2.300   
6   Primary Transit  1SWASP J1407       133.0000               NaN      0.900   
7   Radial Velocity        24 Boo       100.0000            -0.770      0.990   
8   Radial Velocity        24 Sex        72.2084            -0.030      1.540   
9   Radial Velocity        24 Sex        72.2084            -0.030      1.540   
10  Radial Velocity      30 Ari B        39.4000             0.120      1.220   
11  Primary Transit        38 Vir            NaN             0.070      1.180   
12  Radial Velocity         4 Uma        78.5000            -0.250      1.234   
13  Radial Velocity        42 Dra        97.3000            -0.460      0.980   
14  Radial Velocity        47 Uma        13.9700             0.000      1.030   
15  Radial Velocity        47 Uma        13.9700             0.000      1.030   
16  Radial Velocity        47 Uma        13.9700             0.000      1.030   
17  Radial Velocity        51 Peg        14.7000             0.200      1.110   
18  Radial Velocity        55 Cnc        12.3400             0.310      1.015   
19  Radial Velocity        55 Cnc        12.3400             0.310      1.015   

   star_sp_type  
0        G8 III  
1         K4III  
2         K0III  
3          K0 V  
4        G2.5 V  
5         G6III  
6           NaN  
7          G3IV  
8            G5  
9            G5  
10          F6V  
11         F6 V  
12        K1III  
13      K1.5III  
14          G0V  
15          G0V  
16          G0V  
17        G2 IV  
18       K0IV-V  
19       K0IV-V

Take a look at the output from `head(20)` in your notebook. Pay attention to the `eccentricity`, `star_distance`, and `star_metallicity` columns. You'll see some values are listed as `NaN`. `NaN` stands for 'Not a Number' and is the value that `pandas` fills your empty cells with. 

The `NaN`'s have appeared because not every object has data available for every variable. This could be a hassle to deal with, but `pandas` can handle it.  



```{admonition} NaN, null, NA...
:class: myinfo

`NaN` is probably the most common way to show missing data, but you might also come across `NA` or `null`. If you're using some data where the missing values are represented by something else (for example, `99.999` or `----`) you can use the `na_values` argument in `pd.read_csv` to tell it what corresponds to missing data. 
```

We can select all the rows in a dataframe that contain a `NaN` value using the `isna()` function. This is another logical operation; we're selecting the rows that have a `NaN` value. Because we want to select all the rows with a `NaN` and don't care which column it's in, we use `.any(axis=1)`, where `axis=1` says to do the operation on each row. 

In [7]:
exo_df[exo_df.isna().any(axis=1)]

name  orbital_period  semi_major_axis  eccentricity  \
6     1SWASP J1407 b     3725.000000          3.90000           NaN   
11          38 Vir b      825.900000          1.82000         0.030   
30          75 Cet b      691.900000          2.10000           NaN   
34         AD 3116 b        1.982796              NaN         0.146   
35          AD Leo b        2.225990          0.02500         0.030   
...              ...             ...              ...           ...   
3794      beta Pic c     1238.000000          2.72000         0.248   
3798     eps Ind A b    16509.000000         11.55000         0.260   
3807     kepler-80 g       14.645580          0.14000           NaN   
3818       omi UMa b     1630.000000          3.90000           NaN   
3820        pi Men c        6.268340          0.06702           NaN   

      discovered   detection_type     star_name  star_distance  \
6         2012.0  Primary Transit  1SWASP J1407        133.000   
11        2016.0  Primary Transit        38 Vir            NaN   
30        2012.0  Radial Velocity        75 Cet         81.500   
34        2017.0  Primary Transit       AD 3116        186.540   
35        2019.0  Radial Velocity        AD Leo          4.966   
...          ...              ...           ...            ...   
3794      2019.0  Radial Velocity      beta Pic         19.300   
3798      2018.0  Radial Velocity     eps Ind A          3.620   
3807      2017.0  Primary Transit     Kepler-80            NaN   
3818      2012.0  Radial Velocity       omi UMa         56.300   
3820      2018.0  Primary Transit        pi Men         18.370   

      star_metallicity  star_mass star_sp_type  
6                  NaN      0.900          NaN  
11                0.07      1.180         F6 V  
30                0.00      2.490        G3III  
34                 NaN      0.276         M3.9  
35                 NaN        NaN          M4V  
...                ...        ...          ...  
3794               NaN      1.730          A6V  
3798               NaN      0.762          K2V  
3807             -0.56        NaN          NaN  
3818             -0.09      3.090     G4II-III  
3820              0.08      1.094          G0V  

[2866 rows x 11 columns]

At the bottom of the output we can see the size of the dataframe that's returned - 2866 rows by 11 columns. Compare that to the original size of our dataframe (display your dataframe to check the number of rows)... that's a lot of `NaN`s! 

It's handy to check which rows have missing data, but what if we're only concerned with those that aren't missing data? Then we can do the inverse to return only the rows **without** `NaN` values.

We can get this in almost exactly the same command, just one character different. We change
`exo_df.isna()` to `~exo_df.isna()`. The `~` is another way of saying `not`.

In [8]:
exo_df[~exo_df.isna().any(axis=1)]

name  orbital_period  semi_major_axis  eccentricity  discovered  \
0       11 Com b       326.03000           1.2900       0.23100      2008.0   
1       11 UMi b       516.22000           1.5400       0.08000      2009.0   
2       14 And b       185.84000           0.8300       0.00000      2008.0   
3       14 Her b      1773.40000           2.7700       0.36900      2002.0   
4     16 Cyg B b       799.50000           1.6800       0.68900      1996.0   
...          ...             ...              ...           ...         ...   
3827   tau Gem b       305.50000           1.1700       0.03100      2004.0   
3828   ups And b         4.61711           0.0590       0.01186      1996.0   
3829   ups And c       240.93700           0.8610       0.24450      1999.0   
3830   ups And d      1281.43900           2.5500       0.31600      1999.0   
3831   ups And e      3848.86000           5.2456       0.00536      2010.0   

       detection_type star_name  star_distance  star_metallicity  star_mass  \
0     Radial Velocity    11 Com         110.60             -0.35       2.70   
1     Radial Velocity    11 UMi         119.50              0.04       1.80   
2     Radial Velocity    14 And          76.40             -0.24       2.20   
3     Radial Velocity    14 Her          18.10              0.43       0.90   
4     Radial Velocity  16 Cyg B          21.41              0.08       1.01   
...               ...       ...            ...               ...        ...   
3827  Radial Velocity   tau Gem          98.40              0.14       2.30   
3828  Radial Velocity   ups And          13.47              0.09       1.27   
3829  Radial Velocity   ups And          13.47              0.09       1.27   
3830  Radial Velocity   ups And          13.47              0.09       1.27   
3831  Radial Velocity   ups And          13.47              0.09       1.27   

     star_sp_type  
0          G8 III  
1           K4III  
2           K0III  
3            K0 V  
4          G2.5 V  
...           ...  
3827        K2III  
3828         F8 V  
3829         F8 V  
3830         F8 V  
3831         F8 V  

[966 rows x 11 columns]

That's a lot less rows. Oh dear. But what if we don't care about some of the columns with missing data? 

We're only interested in the planet name, period, semi-major axis, detection technique and star mass columns. So let's create a new dataframe that only contains these columns. I'll call this one `planet_df`

First we need to check the names of the columns we want, then we can make our new dataframe using these names:



In [9]:
exo_df.columns

Index(['name', 'orbital_period', 'semi_major_axis', 'eccentricity',
       'discovered', 'detection_type', 'star_name', 'star_distance',
       'star_metallicity', 'star_mass', 'star_sp_type'],
      dtype='object')

In [10]:
planet_df = exo_df[['name', 'orbital_period', 'semi_major_axis', 'detection_type', 'star_mass']]

In [11]:
planet_df

name  orbital_period  semi_major_axis   detection_type  star_mass
0       11 Com b       326.03000           1.2900  Radial Velocity       2.70
1       11 UMi b       516.22000           1.5400  Radial Velocity       1.80
2       14 And b       185.84000           0.8300  Radial Velocity       2.20
3       14 Her b      1773.40000           2.7700  Radial Velocity       0.90
4     16 Cyg B b       799.50000           1.6800  Radial Velocity       1.01
...          ...             ...              ...              ...        ...
3827   tau Gem b       305.50000           1.1700  Radial Velocity       2.30
3828   ups And b         4.61711           0.0590  Radial Velocity       1.27
3829   ups And c       240.93700           0.8610  Radial Velocity       1.27
3830   ups And d      1281.43900           2.5500  Radial Velocity       1.27
3831   ups And e      3848.86000           5.2456  Radial Velocity       1.27

[3832 rows x 5 columns]

Now we can check how many of these rows and columns in `planet_df` contain `NaN` values. Rather than print out the dataframe and read the number underneath, we can use the `shape` attribute.

In [12]:
planet_df[planet_df.isna().any(axis=1)].shape

(1783, 5)

This dataframe has fewer rows with missing data than the first, but still has a lot. We only care about the rows that don't have any missing data. We could use the same operation as we did before to return only the complete rows, but an easier way to do it is to use `dropna`. This removes any rows with missing data. If we'd done this on the original dataframe we'd have lost a lot of rows that are perfectly fine for our analysis but have missing data in another column. By using our new dataframe we don't remove data unnecessarily. Setting `inplace=True` tells Python to edit the dataframe rather than just show us the results. 

In [13]:
planet_df.dropna(inplace=True)


/var/folders/2d/qnv3dwyd2y7dvqhd9dpck1g40000gp/T/ipykernel_13406/2620898407.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  planet_df.dropna(inplace=True)


```{admonition} **SettingWithCopyWarning**
:class: myinfo
You may have got a warning when you ran the `dropna(inplace=True)` command. First - don't panic! This is a warning, not an error. Here the warning is telling us that we're editing a copy of part of another dataframe, not the original dataframe. That's exactly what we intended here so it's not a problem. It becomes more important when you're doing more complex data analysis (more complex that we'll reach in this course), but for now we just need to understand what it's telling us and move on.
```


Let's take another look at our `planet_df` dataframe:

In [14]:
planet_df

name  orbital_period  semi_major_axis   detection_type  star_mass
0       11 Com b       326.03000           1.2900  Radial Velocity       2.70
1       11 UMi b       516.22000           1.5400  Radial Velocity       1.80
2       14 And b       185.84000           0.8300  Radial Velocity       2.20
3       14 Her b      1773.40000           2.7700  Radial Velocity       0.90
4     16 Cyg B b       799.50000           1.6800  Radial Velocity       1.01
...          ...             ...              ...              ...        ...
3827   tau Gem b       305.50000           1.1700  Radial Velocity       2.30
3828   ups And b         4.61711           0.0590  Radial Velocity       1.27
3829   ups And c       240.93700           0.8610  Radial Velocity       1.27
3830   ups And d      1281.43900           2.5500  Radial Velocity       1.27
3831   ups And e      3848.86000           5.2456  Radial Velocity       1.27

[2049 rows x 5 columns]

It definately has fewer rows than before, but we should still check that they're all complete...

(ex-check-data)=
```{admonition} Exercise: Check your data
:class: practice

Use the `isna` function to check that you have properly cleaned the `planet_df` dataframe. Check the shape of the dataframe returned when you use `isna` and its inverse.

[solution](soln_check_data.ipynb)
```

## Key Points
- Missing data is represented as `NaN` or 'Not a Number'
- You can use the `na_values` argument of `pd.read_csv` to tell `pandas` which values correspond to missing data.
- `isna()` can be used to find missing data
- You can use `~` to return the inverse of a logical expression.